flink/v1.12/flink/src/main/java/org/apache/iceberg/flink/sink/BaseDeltaTaskWriter.java

```diff
@@ -83,10 +92,21 @@ public void write(RowData row) throws IOException {
         if (upsert) {
           break;  // UPDATE_BEFORE is not necessary for UPDATE, we do nothing to prevent delete one row twice
         }
-        writer.delete(row);
+        if (deleteSchema != null) {
+          RowData wrap = RowDataProjection.create(schema, deleteSchema).wrap(row);
+          writer.deleteKey(wrap);
+        } else {
+          writer.delete(row);
+        }
```

Comment:
Why does this need to change? I think you just want to fix the `upsert` case.

Reply:
Addtionally, in the `upsert` case, data doesn't come through as `UPDATE_BEFORE`. Though this might be needed to keep CDC data in check.

We've been workin on the PR in my fork but I'll run some tests.